distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/cli/InstallPluginAction.java

```diff
@@ -667,20 +670,47 @@ void verifySignature(final Path zip, final String urlString) throws IOException,
                 throw new IllegalStateException("key id [" + keyId + "] does not match expected key id [" + getPublicKeyId() + "]");
             }
 
+            timer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    terminal.println(
+                        "Downloaded plugin signature verification is taking too long. "
+                            + "This is typically caused by environmental issues on your system related to speed "
+                            + "of random number generation. If you are running on a Linux based system, your "
+                            + "random number generation might be configured to use"
+                            + "/dev/random which blocks until sufficient entropy is achieved on the system. Consider switching to "
+                            + "/dev/urandom instead for faster random number generation."
+                    );
+
+                }
+            }, acceptableSignatureVerificationDelay());
+
             // compute the signature of the downloaded plugin zip
-            final PGPPublicKeyRingCollection collection = new PGPPublicKeyRingCollection(ain, new JcaKeyFingerprintCalculator());
-            final PGPPublicKey key = collection.getPublicKey(signature.getKeyID());
-            signature.init(new JcaPGPContentVerifierBuilderProvider().setProvider(new BouncyCastleFipsProvider()), key);
-            final byte[] buffer = new byte[1024];
-            int read;
-            while ((read = fin.read(buffer)) != -1) {
-                signature.update(buffer, 0, read);
-            }
+            computeSignatureForDownloadedPlugin(fin, ain, signature);
 
             // finally we verify the signature of the downloaded plugin zip matches the expected signature
             if (signature.verify() == false) {
                 throw new IllegalStateException("signature verification for [" + urlString + "] failed");
             }
+        } finally {
+            timer.cancel();
```

Comment:
Why not have this timer be local to computeSignatureForDownload?

Reply:
The main reason I split it this way is to be able to mock computeSignatureForDownload, but I really like your suggestion, there's no reason to pollute the mainline of that code with the timer block. I'll extract into a method that signifies that we are wrapping with a timer.